### ISSUE TITLE
some website elements are inaccessible by keyboard

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#11450]
In the Report a Bug guild, Donald Maness / @ddmaness (4d235d28-bd6f-4859-b222-e9a3f45ea212) has reported that the elements below are inaccessible by keyboard. 
"All Items tested in Chrome on Oct. 16 2019 against WCAG's 2.1 AA standards"
Ref: [Guideline 2.1 - Keyboard Accessible](https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_overview#keyboard-accessible)

I'll leave this as suggestion-discussion for a week or so and then if there's no objections to this approach I'll mark it as help wanted and suggest that a contributor could claim either all these items at once or just one/some of them. If the latter, I'll keep this post updated with the ones that are still open for claiming.

* [ ] The UI element to check off To Do items, Dailies, and Habits:
![image](https://user-images.githubusercontent.com/1495809/67153208-51e95b80-f328-11e9-9f7f-ed76e4b09add.png)

* [ ] The Skills drawer
![image](https://user-images.githubusercontent.com/1495809/67153213-5ca3f080-f328-11e9-9d9d-1463cbf7db66.png)

* [ ] The expandable menu for notifications
![image](https://user-images.githubusercontent.com/1495809/67153214-6594c200-f328-11e9-80b5-7386447bca74.png)

* [ ] Most of the items in the expandable User Icon menu
![image](https://user-images.githubusercontent.com/1495809/67153217-6fb6c080-f328-11e9-941b-5fdae0024a81.png)

* [ ] The dropdown menu for individual tasks
![image](https://user-images.githubusercontent.com/1495809/67153218-780efb80-f328-11e9-9335-ca4c5365bb6a.png)

* [ ] The tag filter controls
![image](https://user-images.githubusercontent.com/1495809/67153219-82c99080-f328-11e9-9f74-1e361efdfd73.png)

----------

Note: tasks accessibility has been handled in #12363

**Comment by ddmaness** [comment]
For some of these examples, non-focusable elements like divs and spans are being used. For example the UI element to check off tasks.

One issue is that these elements are not tab focusable. This can be corrected using the ```tabindex``` attribute. https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex.

The other issue is that these elements are not responsive to keyboard input, only mouse click.  For example, the UI element to check off tasks is, in essence, a checkbox and users that navigate sites by keyboard are accustomed to toggling checkboxes with the ```space``` key.  WAI-ARIA can be used to ensure these elements' behavior is consistent with the behavior to which users are accustomed (https://www.w3.org/TR/wai-aria-practices-1.1/#checkbox)

**Comment by ddmaness** [comment]
Some elements are anchors but are lacking an href.  any anchor tag that does not have an href is not keyboard focusable.  One way around this is to add a dummy href (for example: ```href='#'``` and then, in the event handler, prevent the default link behavior (if I'm not mistaken both ```e.preventDefault()``` and ```return false``` would work)

**Comment by Tressley** [comment]
@ddmaness -- Thank you for all of your feedback surrounding our tab navigation! I will admit that I find this to be a large project as it touches on most UI components on the various pages of Habitica, but I feel like you'd be a great fit for taking this it on. Would you be interested?

**Comment by Alys** [comment]
This accessibility issue was reported yesterday/today by @wslindon (5ccc801c-9c10-41eb-890f-a9ad5c5adee0), so it's probably from a recent change:

"I LOVE the new style changes. Everything looks cleaner and easier to read, e.g. when making a new task with subtasks. Great job Habitica team!
Oh, but Accessibility took a hit (it was never great). **I can Tab into editing subtasks, but can't seem to Tab to the Save button.**"

**Comment by alexandreabeh** [comment]
Hi @Alys ! Looking at the conversation history on this issue, I agree that it's a pretty large project. I'm interested in learning how to make websites more accessible, so this seems like a good opportunity. I went ahead and installed the axe accessibility chrome plugin out of curiosity, and saw a number of things that could be improved beyond keyboard accessibility. The beta version of the tool even lets you select individual components to drill down and target a smaller area at a time.

The UI elements for To-Do, Dailies and Habits are probably most used, so if this is still up for grabs I'd love to start with that. Would you want individual PRs to address just keyboard navigation per UI section, or other screen-reader friendly adds as well? I think I'd have to leave the color contrast ones alone though 😅  

Here are some screenshots to give you an idea of what the chrome installation found.
![inventory accessibility](https://user-images.githubusercontent.com/8680662/86079100-8fb88b00-ba44-11ea-9e11-b99aec79c6a5.PNG)
![main habitica page](https://user-images.githubusercontent.com/8680662/86079102-90e9b800-ba44-11ea-993f-93e82c588b60.PNG)
![modal axe example](https://user-images.githubusercontent.com/8680662/86079104-90e9b800-ba44-11ea-99fb-8ef01da94287.PNG)

**Comment by Alys** [comment]
@alexandreabeh Thanks so much! I've marked this as in progress for you, but for answers to your actual questions, I'll let the staff reply. :)  I'm just a moderator and won't be doing a code review so the staff will have a better idea about the best way to arrange your pull requests.

**Comment by alexandreabeh** [comment]
Ok, sounds good 👍 thanks!

**Comment by Tressley** [comment]
@alexandreabeh -- Thank you for your interest in this! To bring you up to speed, the Habitica team has been making updates to the task modals to address the tab navigation comment made by @Alys. To compliment these changes, I think starting with the task board controls and separating them into individual PRs is a great idea!

**Comment by alexandreabeh** [comment]
Thanks @Tressley ! To make sure I understand what you mean by task board controls (I'm still a relatively new contributor, this is my second issue) - are you referring to the To Do items, Dailies, and Habits in the first screenshot at the top of this thread, plus the dropdown menu on each task?

**Comment by Tressley** [comment]
@alexandreabeh -- Correct! Additionally, I think the tab order should be:

1. Task
2. Task control (+/- for Habits, checkboxes for Dailies/To Do's)
3. Task menu
4. Checklist (if available)

**Comment by alexandreabeh** [comment]
Thanks @Tressley ! I did some research and it sounds like it's a best practice to set tabindex as 0 and let the order be dictated by the natural order of elements in the DOM. https://webaim.org/techniques/keyboard/tabindex#overview

If we follow that, the tab order would be
1. Task control (+, -, or checkbox)
2. Task
3. Task menu
4. Checklist (if available)

What do you think? I could see it being nice if someone was on a screen reader to know about the task before getting to the task control, but I wonder if that should be addressed separately and not through changing the tab order.

**Comment by alexandreabeh** [comment]
Hi @Alys , I just noticed that the linked PR says it will close this issue, but it's only a partial fix. Is there a way to undo whatever automatically triggered that? I didn't want this issue to get lost track of halfway through.

Things the PR addresses out of the original description in this issue:
- to-do items
- dailies
- habits
- dropdown menu for individual tasks

Additional things that are addressed:
- purchasing rewards
- ^ side effect, also fixed purchasing in the shop (market, quests, season, time travelers) 

Filter and toggle menus in the shops still need work and probably should be explicitly added to the list for tracking

**Comment by SabreCat** [comment]
@alexandreabeh I suspect GitHub's parsing doesn't understand that "partially fixes #" is not the same as "fixes #"--I made a small edit to the wording in the PR that looks to have successfully removed the auto-close link!

**Comment by alexandreabeh** [comment]
That would make sense - now I know for next time. Thank you @SabreCat !

**Comment by shanaqui** [comment]
@alexandreabeh Hi! Just checking in whether you still want to work on this further, or whether we should open it up as help wanted again. If you don't want to work on it or we don't hear from you within a week, I'll go ahead! :)

**Comment by Tressley** [comment]
@shanaqui -- This is actively being worked on in #12363

**Comment by alexandreabeh** [comment]
Hi, yes I'm still working on this issue and will be starting another branch for keyboard accessibility in the Shops once the above PR is wrapped up :)

**Comment by paglias** [comment]
I'm setting this issue back to `suggestion discussion` since the tasks accessibility PR by @alexandreabeh has been merged. We can either continue with this issue or split it into new ones
